plugins/out_stackdriver/stackdriver_operation.c

```diff
@@ -64,81 +69,75 @@ int extract_operation(flb_sds_t *operation_id, flb_sds_t *operation_producer,
+        tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
 
-    if (obj->via.map.size != 0) {    	
-        msgpack_object_kv *p = obj->via.map.ptr;
-        msgpack_object_kv *const pend = obj->via.map.ptr + obj->via.map.size;
-
-        for (; p < pend && op_status == NO_OPERATION; ++p) {
-            if (p->val.type == MSGPACK_OBJECT_MAP && p->key.type == MSGPACK_OBJECT_STR
-                && strncmp(OPERATION_FIELD_IN_JSON, p->key.via.str.ptr, p->key.via.str.size) == 0) {
-                
-                op_status = OPERATION_EXISTED;
-                msgpack_object sub_field = p->val;
-                
-                msgpack_object_kv *tmp_p = sub_field.via.map.ptr;
-                msgpack_object_kv *const tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
-
-                /* Validate the subfields of operation */
-                for (; tmp_p < tmp_pend; ++tmp_p) {
-                    if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
-                        continue;
-                    }
-                    if (strncmp("id", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
-                            continue;
-                        }
-                        *operation_id = flb_sds_copy(*operation_id, tmp_p->val.via.str.ptr, tmp_p->val.via.str.size);
-                    }
-                    else if (strncmp("producer", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
-                            continue;
-                        }
-                        *operation_producer = flb_sds_copy(*operation_producer, tmp_p->val.via.str.ptr, tmp_p->val.via.str.size);
-                    }
-                    else if (strncmp("first", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_BOOLEAN) {
-                            continue;
-                        }
-                        if (tmp_p->val.via.boolean) {
-                            *operation_first = FLB_TRUE;
-                        }
-                    }
-                    else if (strncmp("last", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0) {
-                        if (tmp_p->val.type != MSGPACK_OBJECT_BOOLEAN) {
-                            continue;
-                        }
-                        if (tmp_p->val.via.boolean) {
-                            *operation_last = FLB_TRUE;
-                        }
-                    }
-                    else {
-                        /* extra sub-fields */ 
-                        *extra_subfields += 1;
-                    }
-
-                }
+        /* Validate the subfields of operation */
+        for (; tmp_p < tmp_pend; ++tmp_p) {
+            if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
+                continue;
+            }
+
```

Comment:
I have some concerns here, I see that ```extract_operation()``` receives part of the original unpacked msgpack buffer, overriding values can generate corruption or memory leaks. Do we have an unit test for every single ```try_assign_something...()``` here ?

Reply:
I mean: a test that hits ```try_assign_subfield_str()```,  ```try_assign_subfield_bool()```..

Reply:
I have updated the helper functions via your comments.

I did not have unit tests for these functions, but I did not see memory leak after calling `valgrind` now. WDYT?

Reply:
It does receive part of the original unpacked msgpack, but it doesn't change the value of the msgpack. It only changes the value of the received `flb_sds_t`.

Reply:
I understand. Please implement unit tests that matches the sub-functions there.

Reply:
Sorry I just understood your idea. The existed tests can hit the helper functions! @edsiper